### PR TITLE
Moved qos to api.helpers.

### PR DESCRIPTION
**Pull request by k82cn**
**What this PR does / why we need it**:
The `GetPodQoS` is also used by other components, e.g. kube-scheduler and it's not bound to kubelet; moved it to api helpers so client-go.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #N/A

**Release note**:

```release-note-none
```

### CODE REVIEW

**Comment by k82cn** [comment]
xref: #44851

**Comment by k82cn** [comment]
@k8s-bot unit test this

**Comment by k82cn** [comment]
ping :).

**Comment by k82cn** [comment]
ping @jayunit100 / @dchen1107 :).

**Comment by k82cn** [comment]
/cc @lavalamp , @deads2k , @caesarxuchao 

Another example that, the util func `GetPodQOS` is used by several components. I moved it to api.helper; as it's more reasonable to be global helper func instead of kubelet's helper.

**Comment by k82cn** [comment]
ping for this PR :). I'd like to move it to api.helper firstly, and then we can handler it with other help funcs together.

File: pkg/api/v1/helper/helpers_test.go [hunk 1]
```diff
@@ -499,3 +500,155 @@ func TestGetAffinityFromPodAnnotations(t *testing.T) {
 		}
 	}
 }
+
+func TestGetPodQOS(t *testing.T) {
```
Comment (deads2k) [inline pkg/api/v1/helper/helpers_test.go]: I'd like to have a test (I think this one), that does a conversion to internal and runs the internal version of the function to make sure that it agrees with the external version of the function.

**Comment by deads2k** [comment]
I'd like a test to make sure that the implementations stay in sync, but after that, this lgtm.

File: pkg/api/v1/helper/helpers_test.go (hunk 1 above)
Comment (k82cn) [inline pkg/api/v1/helper/helpers_test.go]: sure; I'll add a test for internal one.

Comment (deads2k) [inline pkg/api/v1/helper/helpers_test.go]: > sure; I'll add a test for internal one.

The key piece bit is that its the same test, not a copy/paste.  That way a change in one will cause the test fail.  A separate test won't achieve the same goal.

**Comment by k82cn** [comment]
@deads2k , added a conversion within one test case: https://github.com/kubernetes/kubernetes/pull/44906/commits/ff5e8fba7476412e027b596099400157814b2150#diff-d76ee538bfc38684576a50128a258f11R617

**Comment by caesarxuchao** [comment]
@k82cn are you going to use the helper functions in client-go/pkg/api? They are going to disappear after we move the external types to k8s.io/api.

**Comment by caesarxuchao** [comment]
Can you add a deprecation notice in client-go saying that pkg/api is going to disappear?

**Comment by k82cn** [comment]
@caesarxuchao , this PRs just focus on moving common funcs to helpers. I'd like to follow the way how `k8s.io/api` handle `api.helpers`, including this `GetPodQOS` func.

**Comment by caesarxuchao** [comment]
@k82cn k8s/api only have external types (i.e., excluding pkg/api/types.go), so it won't include any helper functions that in pkg/api/.

Also, client-go will import types from k8s/api and stop copying the entire pkg/api and pkg/apis folder, so client-go won't have the helper functions either.

**Comment by k82cn** [comment]
@caesarxuchao , for where to place those helper funcs, I'd suggest to continue the discuss at https://github.com/kubernetes/kubernetes/issues/44851

In this PRs, it try to avoid introducing package dependency between kubelet & others; after moving to `api.helper`, we just need to discuss how to handle api.helpers.

**Comment by caesarxuchao** [comment]
Ok. Just don't invoke client-go/pkg/api/xxx (i see this PR is not doing this).

Is it possible to add a deprecation notice for client-go/pkg/api?

**Comment by k82cn** [comment]
> Is it possible to add a deprecation notice for client-go/pkg/api?

Done, PTAL :).

File: pkg/api/helper/helpers.go [hunk 2]
```diff
@@ -14,6 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// NOTE: DO NOT use those helper functions through client-go, the
+//       package patch will be changed in the future.
```
Comment (caesarxuchao) [inline pkg/api/helper/helpers.go]: package patch?

Could you remove the indention of the second line?

File: pkg/api/v1/helper/helpers.go [hunk 3]
```diff
@@ -14,21 +14,28 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// NOTE: DO NOT use those helper functions through client-go, the
```
Comment (caesarxuchao) [inline pkg/api/v1/helper/helpers.go]: I haven't made my mind for pkg/api/v1/helper, we might move all the v1 helper function from kubernetes to client-go, so let's only comment in pkg/api/helper

File: pkg/api/helper/helpers.go (hunk 2 above)
Comment (k82cn) [inline pkg/api/helper/helpers.go]: typo, should be "path"; done.

File: pkg/api/v1/helper/helpers.go (hunk 3 above)
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: done.

**Comment by k82cn** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by k82cn** [comment]
@k8s-bot unit test this

**Comment by k82cn** [comment]
@lavalamp , any comments?

**Comment by lavalamp** [comment]
kubelet was clearly the wrong place for this code. I would have preferred to see an api/helpers/qos package, though, instead of adding more to helpers. I'm a bit hesitant to ask for a change like that since you already got some LGTMs, but I think it would be great if you're up for it?

Long term I need to write up how we're going to get out of the current mess where we have to make N versions of every helper function.

**Comment by jayunit100** [comment]
sorry i missed this.... any other comments needed?

**Comment by k82cn** [comment]
>  I'm a bit hesitant to ask for a change like that since you already got some LGTMs, but I think it would be great if you're up for it?

@lavalamp, sure, np. I'll update it. And I'd like to get your comments on how to manage those helper functions in client-go, as this's also come from #44188 .

@jayunit100 , haha, welcome any comments :). maybe also some comments on #44188, I think you're working scheduler refactor (for perf??).

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Comment by k82cn** [comment]
@k8s-bot test this

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Comment by k82cn** [comment]
@k8s-bot bazel test this

**Comment by k82cn** [comment]
@lavalamp , updated according to the comments: add qos sub-directory to helper :).

**Comment by k82cn** [comment]
ping @lavalamp for approve & lgtm :).

**Comment by k82cn** [comment]
@k8s-bot pull-kubernetes-cross test this

**Comment by k82cn** [comment]
@k8s-bot verify test this

**Comment by k82cn** [comment]
The cross build is handling in #45554.

**Comment by k82cn** [comment]
@k8s-bot pull-kubernetes-cross test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by k82cn** [comment]
ping @caesarxuchao for lgtm :).

File: staging/src/k8s.io/client-go/pkg/api/helper/qos/BUILD [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+load(
+    "@io_bazel_rules_go//go:def.bzl",
+    "go_library",
+)
+
+go_library(
+    name = "go_default_library",
+    srcs = ["qos.go"],
+    tags = ["automanaged"],
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/api/helper/qos/BUILD]: Can you add the visibility rule like this: https://github.com/kubernetes/kubernetes/pull/46006

File: staging/src/k8s.io/client-go/pkg/api/install/BUILD [hunk 5]
```diff
@@ -11,7 +11,6 @@ go_library(
     name = "go_default_library",
     srcs = ["install.go"],
     tags = ["automanaged"],
-    visibility = ["//visibility:private"],
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/api/install/BUILD]: Why removing the visibility rule?

Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/api/install/BUILD]: Is it a rebase issue? You removed all the rules.

File: staging/src/k8s.io/client-go/pkg/api/helper/BUILD [hunk 6]
```diff
@@ -8,11 +8,22 @@ load(
...
+        "//vendor/k8s.io/apimachinery/pkg/labels:go_default_library",
+        "//vendor/k8s.io/client-go/pkg/api:go_default_library",
+    ],
+)
+
 go_library(
     name = "go_default_library",
     srcs = ["helpers.go"],
     tags = ["automanaged"],
-    visibility = ["//visibility:private"],
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/pkg/api/helper/BUILD]: Please keep the visibility rule.

File: pkg/api/v1/helper/qos/qos_test.go [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2016 The Kubernetes Authors.
+Copyright 2017 The Kubernetes Authors.
```
Comment (caesarxuchao) [inline pkg/api/v1/helper/qos/qos_test.go]: Please don't change the year.

File: pkg/api/helper/helpers.go [hunk 8]
```diff
@@ -14,6 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// NOTE: DO NOT use those helper functions through client-go, the
```
Comment (caesarxuchao) [inline pkg/api/helper/helpers.go]: We don't need to add the NOTE here now, we restrict the visibility of the package in the BUILD file.

File: staging/src/k8s.io/client-go/pkg/api/helper/BUILD (hunk 6 above)
Comment (k82cn) [inline staging/src/k8s.io/client-go/pkg/api/helper/BUILD]: hm, just run `update-bazel` scripts :). let me check it.

File: pkg/api/v1/helper/qos/qos_test.go (hunk 7 above)
Comment (k82cn) [inline pkg/api/v1/helper/qos/qos_test.go]: ok, got it.

File: pkg/api/helper/helpers.go (hunk 8 above)
Comment (k82cn) [inline pkg/api/helper/helpers.go]: Done

File: pkg/api/v1/helper/qos/qos_test.go (hunk 7 above)
Comment (k82cn) [inline pkg/api/v1/helper/qos/qos_test.go]: done

File: staging/src/k8s.io/client-go/pkg/api/helper/BUILD (hunk 6 above)
Comment (k82cn) [inline staging/src/k8s.io/client-go/pkg/api/helper/BUILD]: done

File: staging/src/k8s.io/client-go/pkg/api/helper/qos/BUILD (hunk 4 above)
Comment (k82cn) [inline staging/src/k8s.io/client-go/pkg/api/helper/qos/BUILD]: done :).

**Comment by k82cn** [comment]
@k8s-bot kops aws e2e test this

**Comment by k82cn** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by caesarxuchao** [comment]
@k82cn Thanks for bearing with me :)

**Comment by k82cn** [comment]
my pleasure :).

**Pull request merged**